app/email-incoming/index.ts

```diff
@@ -0,0 +1,180 @@
+/*!
+ * Copyright © 2022 United States Government as represented by the Administrator
+ * of the National Aeronautics and Space Administration. No copyright is claimed
+ * in the United States under Title 17, U.S. Code. All Other Rights Reserved.
+ *
+ * SPDX-License-Identifier: NASA-1.3
+ */
+
+import {
+  ListUsersInGroupCommand,
+  CognitoIdentityProviderClient,
+} from '@aws-sdk/client-cognito-identity-provider'
+import { S3Client, GetObjectCommand } from '@aws-sdk/client-s3'
+import { DynamoDBAutoIncrement } from '@nasa-gcn/dynamodb-autoincrement'
+import { DynamoDBClient } from '@aws-sdk/client-dynamodb'
+import { DynamoDBDocument } from '@aws-sdk/lib-dynamodb'
+import { SESClient, SendEmailCommand } from '@aws-sdk/client-ses'
+import type { S3Event } from 'aws-lambda'
+
+import type { Source } from 'mailparser'
+import { simpleParser } from 'mailparser'
+
+import {
+  emailIsAutoReply,
+  subjectIsValid,
+  formatAuthor,
+} from '../routes/circulars/circulars.lib'
+
+import { tables } from '@architect/functions'
+import { extractAttributeRequired, extractAttribute } from '~/lib/cognito'
+
+const s3 = new S3Client({})
+const client = new CognitoIdentityProviderClient({})
+const sesClient = new SESClient({})
+const doc = DynamoDBDocument.from(new DynamoDBClient({}))
+
+const USER_POOL_ID = process.env.USER_POOL_ID
+
+const domain =
+  process.env.NODE_ENV === 'development'
+    ? 'dev.'
+    : process.env.NODE_ENV === 'test'
+    ? 'test.'
+    : ''
+
+export async function handler(event: S3Event) {
+  const getObjectParamPromises = event.Records.filter(
+    (record) => record.eventName == 'ObjectCreated:Put'
+  ).map((record) => {
+    return s3.send(
+      new GetObjectCommand({
+        Bucket: record.s3.bucket.name,
+        Key: decodeURIComponent(record.s3.object.key.replace(/\+/g, ' ')),
+      })
+    )
+  })
+
+  // Get data from S3
+  const objects = await Promise.all(getObjectParamPromises)
+
+  objects.forEach(async (object) => {
+    if (!object.Body) throw new Error('S3 object has no body')
+
+    const parsed = await simpleParser(object.Body as Source)
```

Comment:
Why is Typescript unable to infer that `object.Body` is satisfies the argument type of `simpleParser`?

Reply:
This is the error without the cast:

Argument of type 'SdkStream<Readable | ReadableStream<any> | Blob | undefined>' is not assignable to parameter of type 'Source'.
  Type 'ReadableStream<any> & SdkStreamMixin' is not assignable to type 'Source'.ts(2345)

Reply:
Are they actually compatible types? Does this code work?

Reply:
It did, but all of the testing was done when it was the version I uploaded manually (before it was in this project). I will try running the esbuild, and manually putting the out file in the lambda to verify